lib/plugins/aws/package/compile/events/apiGateway/lib/hack/updateStage.js

```diff
@@ -198,32 +202,43 @@ function handleLogs() {
 
 function handleTags() {
   const provider = this.state.service.provider;
-  const tagsMerged = Object.assign({}, provider.stackTags, provider.tags);
-
-  const tagsToCreate = _.entriesIn(tagsMerged).map(pair => ({
-    key: String(pair[0]),
-    value: String(pair[1]),
-  }));
-  if (tagsToCreate) {
-    tagsToCreate.forEach((tag) => {
-      const operation = { op: 'replace', path: `/variables/${tag.key}`, value: tag.value };
-      this.apiGatewayStagePatchOperations.push(operation);
+  const tagsMerged = _.mapValues(Object.assign({}, provider.stackTags, provider.tags),
+      v => String(v));
+  const currentTags = this.apiGatewayStageState.tags || {};
+  const tagKeysToBeRemoved = Object.keys(currentTags)
+    .filter(currentKey => !_.isString(tagsMerged[currentKey]));
+
+  const restApiId = this.apiGatewayRestApiId;
+  const stageName = this.options.stage;
+  const region = this.options.region;
+  const resourceArn = `arn:aws:apigateway:${region}::/restapis/${restApiId}/stages/${stageName}`;
+
+  if (tagKeysToBeRemoved.length > 0) {
+    this.apiGatewayUntagResourceParams.push({
+      resourceArn,
+      tagKeys: tagKeysToBeRemoved,
     });
   }
-
-  if (this.apiGatewayStageState.variables) {
-    const stateTagKeys = Object.keys(this.apiGatewayStageState.variables);
-    const newTagKeys = Object.keys(tagsMerged);
-    const tagsToRemove = _.difference(stateTagKeys, newTagKeys);
-    if (tagsToRemove) {
-      tagsToRemove.forEach((tagKey) => {
-        const operation = { op: 'remove', path: `/variables/${tagKey}` };
-        this.apiGatewayStagePatchOperations.push(operation);
-      });
-    }
+  if (!_.isEqual(currentTags, tagsMerged) && _.size(tagsMerged) > 0) {
+    this.apiGatewayTagResourceParams.push({
+      resourceArn,
+      tags: tagsMerged,
```

Comment:
Why not difference of `tagsMerged` against `currentTags` ?

Reply:
Because tagResouce API can ignore unchanged keys, so no need to filter out unchanged keys from `tagsMerged`, as far as I tested.

Reply:
Ok, indeed no harm in that